src/io/mod.rs

```diff
@@ -0,0 +1,202 @@
+pub use file_reader::{LocalFsReaderConfig, ReaderConfiguration};
+
+mod file_reader {
+    use std::fs;
+    use std::io::prelude::*;
+    use std::io::{BufReader, Result};
+    use std::path::{Path, PathBuf};
+
+    pub trait ReaderConfiguration {}
+
+    pub struct LocalFsReaderConfig {
+        filter_ext: Option<std::ffi::OsString>,
+        expect_dir: bool,
+        dir_path: PathBuf,
+    }
+
+    impl LocalFsReaderConfig {
+        /// Read all the files from a directory or a path.
+        pub fn new<T: Into<PathBuf>>(path: T) -> LocalFsReaderConfig {
+            LocalFsReaderConfig {
+                filter_ext: None,
+                expect_dir: true,
+                dir_path: path.into(),
+            }
+        }
+        /// Only will read files with a given extension.
+        pub fn filter_extension<T: Into<String>>(&mut self, extension: T) {
+            self.filter_ext = Some(extension.into().into());
+        }
+        /// Default behaviour is to expect the directory to exist in every node,
+        /// if it doesn't the executor will panic.
+        pub fn expect_directory(&mut self, should_exist: bool) {
+            self.expect_dir = should_exist;
+        }
+
+        fn build(self) -> LocalFsReader {
+            LocalFsReader::new(self)
+        }
+    }
+
+    impl ReaderConfiguration for LocalFsReaderConfig {}
+
+    /// Reads all files specified in a given directory from the local directory
+    /// on all executors on every worker node.
+    struct LocalFsReader {
+        path: PathBuf,
+        is_single_file: bool,
+        filter_ext: Option<std::ffi::OsString>,
+        expect_dir: bool,
+        files: Vec<Vec<PathBuf>>,
+        open_file: Option<BufReader<fs::File>>,
```

Comment:
Only exact byte reads are done on `open_file`, so it can just be `Option<fs::File>`